neqo-transport/src/ecn.rs

```diff
@@ -38,73 +27,34 @@ enum EcnValidationState {
     Capable,
 }
 
-#[derive(Debug)]
+#[derive(Debug, Default)]
 pub struct EcnInfo {
     /// The current state of ECN validation on this path.
     state: EcnValidationState,
-
+    /// The number of packets sent so far on the path during the ECN validation.
+    sent: usize,
     /// The ECN counts received in the last ACK on this path, for each packet number space.
-    /// Won't be updated after ECN has been disabled on a path.
+    /// Won't be updated after ECN has been tested on the path.
     count: EnumMap<PacketNumberSpace, EcnCount>,
 }
 
 impl EcnInfo {
-    pub fn new(now: Instant) -> Self {
-        Self {
-            state: EcnValidationState::Testing {
-                start: now,
-                count: 0,
-                lost: 0,
-            },
-            count: EnumMap::default(),
-        }
-    }
-
+    /// Count the number of packets sent out on this path during ECN validation.
+    /// Exit ECN validation if the number of packets sent exceeds `ECN_TEST_COUNT`.
+    /// We do not implement the part of the RFC that says to exit ECN validation if the time since
+    /// the start of ECN validation exceeds 3 * PTO, since this seems to happen much too quickly.
     pub fn count_packets_out(&mut self) {
```

Comment:
This doesn't seem to do any counting.  It adds one more to the count.  I'd use `on_packet_sent` or `track_sent_packet`.

That said, I think that your validation is going to be a bit challenging if you have multiple paths.  We don't really get clear signals about which path the ECN markings apply to, so there are a lot of scenarios to think through.

1. The existing path is happily using ECN and a new path comes online.  On that path, we start testing and send marked packets.  Those packets will be acknowledged along with all the packets on the original path.  This should work fine because both paths pass the markings correctly.
2. The new path is bleaching, but the original is fine.  Here, I see no problem either.  Correct markings on the original path will not prevent the gap from being noticed.
3. The old path is bleaching, but the new one is fine.  This is a problematic case.  We aren't able to attribute a lack of markings to either path.
4. Both are bleaching.  This is also fine, in the sense that ECN is disabled on the new path.

Maybe case 3 is just something we can't handle, but I don't think that's true.  We have a count of the probing packets we sent during the testing phase, so we should be able to look for an increase in counts that match.  What we don't have is anything that allows us to track which of the acknowledged packets were sent on the new path.  We could use ACK tokens for that, I guess.

We could also decide that we don't care about this case.  Bleaching can be infectious.  Though that's not particularly satisfactory.  We would leak that we previously used a bleaching path, which isn't great.

Either way, I'd like to see a few tests that covered these cases.